src/lib/pubkey/hss_lms/lms.cpp

```diff
@@ -0,0 +1,351 @@
+      throw Decoding_Error("To few signature bytes while parsing LMS signature.");
+   }
+
+   // Alg. 6a 2.j.
+   auto auth_path = slicer.take<LMS_AuthenticationPath>(lms_params.m() * lms_params.h());
+
+   return LMS_Signature(q, std::move(lmots_sig), lms_type, LMS_AuthenticationPath(auth_path));
+}
+
+LMS_PublicKey::LMS_PublicKey(const LMS_PrivateKey& sk) : LMS_Instance(sk), m_lms_root(sk.lms_params().m()) {
+   lms_treehash(StrongSpan<LMS_Tree_Node>(m_lms_root), std::nullopt, std::nullopt, sk);
+}
+
+bool LMS_PublicKey::verify_signature(const LMS_Message& msg, const LMS_Signature& sig) const {
+   if(lms_root().size() != lms_params().m()) {
+      // LMS public key (T[1] part) has unexpected length
+      return false;
+   }
+   if(lms_params().algorithm_type() != sig.lms_type()) {
+      // LMS algorithm type does not match with the signature's
+      return false;
+   }
+   // Alg. 6a 2.g.
+   if(lmots_params().algorithm_type() != sig.lmots_sig().algorithm_type()) {
+      // LMOTS algorithm type does not match with the signature's
+      return false;
+   }
+   // Alg. 6a 2.i.
+   if(sig.q() >= (1ULL << uint64_t(lms_params().h()))) {
+      return false;
+   }
+   // Alg 6. 3.
+   std::optional<LMS_Tree_Node> Tc = lms_compute_root_from_sig(msg, sig);
+   if(!Tc.has_value()) {
+      return false;
+   }
+   // Alg 6. 4.
+   return Tc.value() == lms_root();
+}
+
+std::optional<LMS_Tree_Node> LMS_PublicKey::lms_compute_root_from_sig(const LMS_Message& msg,
+                                                                      const LMS_Signature& sig) const {
+   // Alg. 6a 2.c, 2.g
+   if(lms_params().algorithm_type() != sig.lms_type() ||
+      lmots_params().algorithm_type() != sig.lmots_sig().algorithm_type()) {
+      return std::nullopt;
+   }
+
+   const LMS_Params lms_params = LMS_Params::create_or_throw(sig.lms_type());
+   const LMOTS_Signature& lmots_sig = sig.lmots_sig();
+   const LMOTS_Params lmots_params = LMOTS_Params::create_or_throw(lmots_sig.algorithm_type());
+   const LMOTS_K Kc = lmots_compute_pubkey_from_sig(lmots_sig, msg, identifier(), sig.q());
+   const auto hash = HashFunction::create_or_throw(lms_params.hash_name());
+
+   auto hash_pair_func = get_hash_pair_func_for_identifier(lms_params, identifier());
+
+   auto lms_address = TreeAddress(lms_params.h());
+   lms_address.set_address(LMS_TreeLayerIndex(0), LMS_Tree_Node_Idx(sig.q().get()));
+
+   LMOTS_Public_Key pk_candidate(lmots_params, identifier(), sig.q(), Kc);
+   LMS_Tree_Node tmp(lms_params.m());
+   lms_gen_leaf(tmp, pk_candidate, lms_address, *hash);
+
+   LMS_Tree_Node root(lms_params.m());
+
+   compute_root(StrongSpan<LMS_Tree_Node>(root),
+                sig.auth_path(),
+                sig.q(),
+                StrongSpan<const LMS_Tree_Node>(tmp),
+                lms_params.m(),
+                LMS_TreeLayerIndex(lms_params.h()),
+                0,
+                hash_pair_func,
```

Comment:
As stated above: When these are lambdas that have a closured `unique_ptr<>` they won't be able to be copied.

```suggestion
                std::move(hash_pair_func),
```